### ISSUE TITLE
Refactor how InputMedia* subclasses handle inits

### ISSUE THREAD

**Issue by Bibo-Joshi** [python-telegram-bot/python-telegram-bot#2573]
Currently all the subclasses handle the init completely on their own, while it would make more sense to move all attributes shared between the subclasses to `InputMedia`, i.e. call `super().__init__` like we do for `InlineQueryResult*`. This includes the attributes `caption_entities`, `parse_mode`, `type`, `media` & `caption`

See https://t.me/c/1494805131/16382 + replies for discussion.

**Comment by eldbud** [comment]
i'm on it

**Comment by eldbud** [comment]
question regrading ```caption_entities``` - if i understand correctly, the ```InputMedia.caption_entities``` class variable is never actually used, because each subclass instantiates an instance variable under the same name, which is the one that is actually being accessed, so i believe ```InputMedia.caption_entities``` is in fact redundant. am i correct, or am i missing some usage of the this base class variable?

**Comment by eldbud** [comment]
regarding the ```type``` attribute you included in the description - i don't see it anywhere in ```InputMedia``` or its inheritors. did you mean something else? maybe ```thumb```?

**Comment by Bibo-Joshi** [comment]
IIRC we added `caption_entities` to `InputMedia` so that `InputMedia.to_dict` would be guaranteed to work. But yes, this should be refactored to an instance variable that's set in `InputMedia.__init__`.

**Comment by eldbud** [comment]
Why are the ```Video```/```Animation```/```Audio```/```Document```/```PhotoSize``` classes not inheritors of the ```File``` class?
If there is a reason, then shouldn't they still all be inheriting from a common base class? they all have many identical attributes as well.

**Comment by eldbud** [comment]
@Bibo-Joshi This is pretty much the EXACT implementation i have been working on since @Poolitzer 's comment.
There's a ```_medium.py``` module which contains the implementation of the new ```Medium``` class. 
Of course not part of the API, merely a design improvement.

**Comment by Bibo-Joshi** [comment]
Right, I forgot about `PhotoSize` not having a thumb. Still, it should be possible to move `de_json` to the base class by including a  check a lá `if 'thumb' in data:`.
You'll have to change how the classes in `sticker.py` override `de_json`, then. Actually, if you do that anyway, you can just remove `MaskPosition.de_json`, as it doesn't do anything that `TelegramObject.de_json` does not do already.

Regarding the namings: I think the class name should be private rather then the module. Also IMO naming it e.g. `_BaseMedium` or `_MediumObject` makes it more clear that it's supposed to be used as base class for something …

**Comment by Poolitzer** [comment]
Im good with that implementation jup

**Comment by eldbud** [comment]
regarding module/class as private - i made the module private, because that way already on import it will raise a warning the a protected module is being imported.
By naming the class as private, only when importing that class from the module the warning will be raised, which normally happens later while coding. just my two cents.

**Comment by Bibo-Joshi** [comment]
I see your point. However, none of the existing modules are marked private even though the module structure is not considered to be part of the public API. *If* we want to change that, I suggest to do that coordinated in a single PR. For now, I'd vote to stick to the existing convention :)

**Comment by Bibo-Joshi** [comment]
FYI: I've just openend #2680 ;)

**Comment by eldbud** [comment]
oh i got my eyes on it alright. if no one will take it by then, i may give it a try after completing the current task.
Glad you liked the protected module approach!

**Comment by eldbud** [comment]
question regarding usage of ```parse_file_input()``` in the ```inputfile``` module classes:

I see that in some of the subclasses the ```tg_type``` is provided, and in others it is not. For example in ```InputMediaPhoto``` and ```InputMediaDocument it provides the ```PhotoSize``` and ```Document``` class correspondingly, but in ```InputMediaAnimation``` ```InputMediaAudio``` and ```InputMediaVideo``` no ```tg_type``` is provided. 
Is this intentional? if so, why so?

**Comment by Bibo-Joshi** [comment]
IISC in the cases where it's not provided, this is already handled in another `if` clause, e.g. for `InputMediaAnimation` the first if-clause handles the case of `Animation`-type input:

https://github.com/python-telegram-bot/python-telegram-bot/blob/c3e3bb77e507f1997c1eb66f81265da4da12b68b/telegram/files/inputmedia.py#L143-L149

the point is here that we can extract additional information from the TG-Type-input.

**Comment by eldbud** [comment]
so as i mentioned in the Dataclasses discussion, i tried using mixin classes to refrain from repeating ```width``` ```height``` ```duration``` ```thumb```  ```file_name``` ```mime_type``` ```title``` over and over. 
Planned on showing it to you after running tests.
I ran tests.
I assume you will not be surprised to hear it did not fare well.
So i will keep the boilerplate code, so it seems.

**Comment by Bibo-Joshi** [comment]
Okay, I understand the idea 👍
There *are* some ways of making slots working with multiple inheritance (this [SO answer](https://stackoverflow.com/a/28059785/10606962) is a very nice write up). But my gut tells that implementing workarounds for this probably is more complicated than adding a few lines of boilerplate :D If we go for dataclasses, this would even reduce the boiler plate lines.
I suggest to stick to boiler plate for now and keep the idea in mind :) It's nice to see how you are digging into the library, though! 😃

**Comment by eldbud** [comment]
yep, boilerplate it will remain.
I did start investigating about how maybe i will after all be able to incorporate multiple inheritance with slots; and then i found the SO link you shared, which lead to the understanding that incorporating this workaround will be too big of a weapon for the needs at hand.

**Comment by Poolitzer** [comment]
Since we are giving out compliments here, let me agree to that one: I noticed the PRs from you, well done. Its a great feeling if you get a new face in an OS project because they want to contribute. Thank you very much.

**Comment by eldbud** [comment]
going back to the subject, i'm still having some ```__slots__``` issues:
When calling the ```de_json``` method of the ```_BaseMedium``` class, for example when initializing an ```Audio``` object, i am getting this error, upon the ```return cls(bot=bot, **data)``` line:
```TypeError: __init__() missing 1 required positional argument: 'duration'```

I checked, and the ```data``` dictionary does include the ```duration``` key. So as i understand it, the reason this exception arises, is that ```_BaseMedium.__init__()``` order of arguments is different from that of ```Audio```. I am unable to change the order, because the ```duration``` argument is mandatory and has no default value, whereas arguments which go to ```_BaseMedium```'s ```__init__()``` (e.g., ```file_size```) and should supposedly come before the subclass's attributes, is initialized with ```None```, and thus cannot be moved to be before the subclass's attributes (such as the ```duration``` example i brought here.

I hope the articulation above is not too confusing...

Any suggestions as to how to approach this?

**Comment by Bibo-Joshi** [comment]
I'm not sure if I can follow. Why do you want to call `_BaseMedium.de_json`? shouldn't you rather call `Audio.de_json`, which calls `Audio.__init__(**data, bot=bot)`? In `Audio.__init__`, you can call `super().__init__(file_id=file_id, …)` and by using kwargs the order shouldn't matter …

**Comment by eldbud** [comment]
i'm not calling either, ```de_json``` calls ```cls(bot, **data)```.
But yes, i was missing the naming of the arguments in ```super().__init__()```. i believe that should fix it. thanks!

**Comment by Bibo-Joshi** [comment]
Mh, the author of #2700 has a similar situation … You can commit with `git commit … --no-verify` to skip the `pre-commit` step. THen we'll have a look at the automated tests and see how they behave :)

**Comment by Bibo-Joshi** [comment]
Sounds like something went wrong with that submodule … Try running `$ git submodule update --init --recursive`. If you keep having problems, feel free to ping me on TG (@BiboJoshi)

**Comment by eldbud** [comment]
if it helps in any way, this is the failing line in ```check_shortcut_signature```:
```if isinstance(bot_sig.parameters[kwarg].annotation, type):``` (conftest.py, line 445).
And upon checking what is the type - ```<class 'typing._UnionGenericAlias'>```
